nifi-nar-bundles/nifi-standard-bundle/nifi-standard-processors/src/main/java/org/apache/nifi/processors/standard/DetectDuplicateRecord.java

```diff
@@ -0,0 +1,620 @@
+                    logger.debug("Removal of expired cached entry with key {} returned {}", new Object[]{cacheKey, status});
+                } else {
+                    serializableFilter = cacheValue.getFilter();
+                }
+            }
+
+            final FilterWrapper filter = FilterWrapper.create(serializableFilter);
+
+            final String recordHashingAlgorithm = context.getProperty(RECORD_HASHING_ALGORITHM).getValue();
+            final MessageDigest messageDigest = DigestUtils.getDigest(recordHashingAlgorithm);
+            final Boolean matchWholeRecord = context.getProperties().keySet().stream().noneMatch(p -> p.isDynamic());
+
+            final RecordReaderFactory readerFactory = context.getProperty(RECORD_READER).asControllerService(RecordReaderFactory.class);
+            final RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+            final RecordReader reader = readerFactory.createRecordReader(flowFile.getAttributes(), session.read(flowFile), logger);
+
+            FlowFile nonDuplicatesFlowFile = session.create(flowFile);
+            final FlowFile duplicatesFlowFile = session.create(flowFile);
+            final RecordSchema writeSchema = writerFactory.getSchema(flowFile.getAttributes(), reader.getSchema());
+            final RecordSetWriter nonDuplicatesWriter = writerFactory.createWriter(getLogger(), writeSchema, session.write(nonDuplicatesFlowFile));
+            final RecordSetWriter duplicatesWriter = writerFactory.createWriter(getLogger(), writeSchema, session.write(duplicatesFlowFile));
+
+            nonDuplicatesWriter.beginRecordSet();
+            duplicatesWriter.beginRecordSet();
+            Record record;
+
+            while ((record = reader.nextRecord()) != null) {
+                String recordValue;
+
+                if (matchWholeRecord) {
+                    recordValue = record.getSerializedForm().get().getSerialized().toString();
+                } else {
+                    final List<String> fieldValues = new ArrayList<>();
+                    for (final String recordPathText : recordPaths) {
+                        final PropertyValue recordPathPropertyValue = context.getProperty(recordPathText);
+                        final RecordPath recordPath = recordPathCache.getCompiled(recordPathText);
+                        final RecordPathResult result = recordPath.evaluate(record);
+                        fieldValues.addAll(result.getSelectedFields()
+                                        .map(f -> recordPathPropertyValue.evaluateAttributeExpressions(flowFile).getValue())
+                                        .collect(toList())
+                        );
+                    }
+                    recordValue = Joiner.on('~').join(fieldValues);
+                }
+
+                final String recordHash = Hex.encodeHexString(messageDigest.digest(getBytesUtf8(recordValue)));
```

Comment:
I think we already have commons-codec's `DigestUtils` class available if you would rather use that.

Reply:
I am using that up above on line 417. Since the hashing algorithm is a dynamic selection, there was not a more convenient interface exposed for this so I opted to initialize the message digest ahead of time before entering the loop by using:

```
final String recordHashingAlgorithm = context.getProperty(RECORD_HASHING_ALGORITHM).getValue();
final MessageDigest messageDigest = DigestUtils.getDigest(recordHashingAlgorithm);
```

If I did it the other way, I would have to write if/else statements inside the loop using interfaces similar to `DigestUtils.sha1Hex("AppendedRecordPathValues")` for each supported algorithm, causing the algorithm check to occur on each iteration of the loop. If there is a cleaner call to `MessageDigest` I am open to suggestions.